src/node/event_stream.rs

```diff
@@ -0,0 +1,132 @@
+                    incoming_msgs.remote_addr()
+                );
+                Self::spawn_messages_handler(stage.clone(), incoming_msgs, xorname)
+            }
+        });
+    }
+
+    // Spawns a task which handles each new incoming message from a connection with a peer
+    fn spawn_messages_handler(
+        stage: Arc<Mutex<Stage>>,
+        mut incoming_msgs: IncomingMessages,
+        xorname: XorName,
+    ) {
+        let _ = tokio::spawn(async move {
+            while let Some(msg) = incoming_msgs.next().await {
+                match msg {
+                    QuicP2pMsg::UniStream { bytes, src, .. } => {
+                        trace!(
+                            "New message ({} bytes) received on a uni-stream from: {}",
+                            bytes.len(),
+                            src
+                        );
+                        // Since it's arriving on a uni-stream we treat it as a Node
+                        // message which we need to be processed by us, as well as
+                        // reported to the event stream consumer.
+                        spawn_node_message_handler(stage.clone(), bytes, src);
+                    }
+                    QuicP2pMsg::BiStream {
+                        bytes,
+                        src,
+                        send,
+                        recv,
+                    } => {
+                        trace!(
+                            "New message ({} bytes) received on a bi-stream from: {}",
+                            bytes.len(),
+                            src
+                        );
+
+                        // Since it's arriving on a bi-stream we treat it as a Client
+                        // message which we report directly to the event stream consumer
+                        // without doing any intermediate processing.
+                        let event = Event::ClientMessageReceived {
+                            content: bytes,
+                            src,
+                            dst: DstLocation::Node(xorname),
+                            send,
+                            recv,
+                        };
+
+                        stage.lock().await.send_event(event).await;
+                    }
+                }
+            }
+        });
+    }
+}
+
+fn spawn_node_message_handler(stage: Arc<Mutex<Stage>>, msg_bytes: Bytes, sender: SocketAddr) {
+    let _ = tokio::spawn(async move {
+        match Message::from_bytes(&msg_bytes) {
+            Err(error) => {
+                debug!("Failed to deserialize message: {:?}", error);
+            }
+            Ok(msg) => {
+                trace!("try handle message {:?}", msg);
+                // Process the message according to our stage
+                if let Err(err) = stage
+                    .lock()
+                    .await
+                    .process_message(sender, msg.clone())
+                    .await
```

Comment:
Yeah, I was thinking about something similar, but not necessarily tied to CRDTs. Something like this:
- Keep the node state in a `RwLock` instead of `Mutex`, to allow multiple concurrent readers
- Each message handler is passed only immutable reference to the state (with `RwLock::read`)
- Instead of mutating the state, the handlers would push an `Op` into the op queue/channel
- There is going to be one dedicated task which will keep popping from the op queue and applying the op to the state. This is the only task that will have writable access to the state (using `RwLock::write`)
- We can also use this mechanism as sort of general purpose async notification system. So for example, when a message handler `tokio::spawn`s a task and that task needs to pass some info back to the system, it can also push it into the op queue. Some examples of where this would be useful is the DKG (which can run in its own task) and the message resending which needs some way to notify about failed peers.

How does that sound?